### PR TITLE
CNAO, KubevirtIpamController: Support component by a feature gate

### PR DESCRIPTION
**Pull request by oshoval**
**What this PR does / why we need it**:

`KubevirtIpamController` allows having persistent IPs for Kubevirt VM user defined networks.
OVN k8s implements the SDN side, hence required to be deployed as well.

Since the feature is alpha, this PR presents a feature gate flag, allows
installing the component via CNAO (default value is false).
The feature is also feature gated on OCP itself [1].

Note that if used on K8s cluster (non OCP), a cert manager is required to be deployed,
cert-manager is recommended.

Feature design https://github.com/openshift/enhancements/pull/1456

[1] https://github.com/openshift/api/pull/1913

**Reviewer Checklist**

> Reviewers are supposed to review the PR for every aspect below one by one. To check an item means the PR is either "OK" or "Not Applicable" in terms of that item. All items are supposed to be checked before merging a PR. 

- [ ] PR Message
- [ ] Commit Messages
- [ ] How to test
- [ ] Unit Tests
- [ ] Functional Tests
- [ ] User Documentation
- [ ] Developer Documentation
- [ ] Upgrade Scenario
- [ ] Uninstallation Scenario
- [ ] Backward Compatibility
- [ ] Troubleshooting Friendly

**Jira Ticket**:

```jira-ticket
https://issues.redhat.com/browse/CNV-42721
```

**Release note**:

```release-note
Support deploying KubevirtIpamController via a feature gate
```

### CODE REVIEW

File: vendor/github.com/kubevirt/cluster-network-addons-operator/pkg/apis/networkaddonsoperator/shared/networkaddonsconfig_types.go
Comment (oshoval) [inline vendor/github.com/kubevirt/cluster-network-addons-operator/pkg/apis/networkaddonsoperator/shared/networkaddonsconfig_types.go:20]: CNAO need to be vendored once there is a new release, and then this file wont be needed
until then this PR is on hold

File: controllers/operands/networkAddons_test.go
Comment (nunnatsa) [inline controllers/operands/networkAddons_test.go:661]: please use ginkgo's context; e.g.
```suggestion
		kubevirtIpamTester := func(ctx context.Context, o ipamAnnotationParams) {
			existingCNAO, err := NewNetworkAddons(hco)
			Expect(err).ToNot(HaveOccurred())
			if o.ipamExists {
				existingCNAO.Spec.KubevirtIpamController = &networkaddonsshared.KubevirtIpamController{}
			}

			if o.setFeatureGate {
				hco.Spec.FeatureGates.DeployKubevirtIpamController = ptr.To(o.featureGateValue)
			}

			cl := commontestutils.InitClient([]client.Object{hco, existingCNAO})
			handler := (*genericOperand)(newCnaHandler(cl, commontestutils.GetScheme()))
			res := handler.ensure(req)
			Expect(res.UpgradeDone).To(BeFalse())
			Expect(res.Err).ToNot(HaveOccurred())

			foundCNAO := &networkaddonsv1.NetworkAddonsConfig{}
			Expect(
				cl.Get(ctx,
```

File: controllers/operands/networkAddons_test.go
Comment (nunnatsa) [inline controllers/operands/networkAddons_test.go:651]: I think we can do it with one boolean: just pass the required FG value to the test, and set it unconditionally.

Comment (oshoval) [reply controllers/operands/networkAddons_test.go:651]: but then it means we set even if false, overriding the default value
it can work of course, i just copy pasted the from the other tests

File: api/v1beta1/hyperconverged_types.go
Comment (nunnatsa) [inline api/v1beta1/hyperconverged_types.go:442]: Please add some more details here. What's KubevirtIpamController? even add a reference.

Comment (oshoval) [reply api/v1beta1/hyperconverged_types.go:442]: we have it here
docs/cluster-configuration.md
and i just followed the other structs, do you still think it is required?

**Review by nunnatsa — COMMENTED** [review]
Thanks for the PR @oshoval 

I think it's premature. Let's wait for CANO release. 

Added a few comment inline. 

Please also add more info in the PR description. This description should convince the reader that a new API is required, and explain the context, the requirements and so on.

File: controllers/operands/networkAddons_test.go
Comment (nunnatsa) [reply controllers/operands/networkAddons_test.go:651]: I won't hold the PR for that. your call

File: controllers/operands/networkAddons_test.go
Comment (nunnatsa) [inline controllers/operands/networkAddons_test.go:642]: please inline in the DescribeTable that uses this function. I can't see the benefit of this function, and even less benefit of assign it to a variable. I find it more readable to inline the logic in the Describe table itself.

Please also do the same for `ksdTester`

File: api/v1beta1/hyperconverged_types.go
Comment (nunnatsa) [reply api/v1beta1/hyperconverged_types.go:442]: Still, we need it in the PR description. If we missed it in previous PR, that too bad, but we do need it

File: controllers/operands/networkAddons_test.go
Comment (oshoval) [reply controllers/operands/networkAddons_test.go:642]: `ksdTester` is used twice, so we can't do that there as straight forward as here
and i think any change to `ksdTester` deserves please a separate PR
it is totally orthogonal to this PR
having a topic centric PRs help later when need references and such
is it ok ?

File: api/v1beta1/hyperconverged_types.go
Comment (nunnatsa) [reply api/v1beta1/hyperconverged_types.go:442]: We don't need it to be too long, but this is an input to the CRD, and part of the API. it should be very clear to the user. the docs/cluster-configuration.md is for human readable text.

File: controllers/operands/networkAddons_test.go
Comment (oshoval) [reply controllers/operands/networkAddons_test.go:651]: we even test it "should not have KIC if feature gate does not exist"
i prefer to keep as is

File: deploy/olm-catalog/community-kubevirt-hyperconverged/1.13.0/manifests/kubevirt-hyperconverged-operator.v1.13.0.clusterserviceversion.yaml
Comment (oshoval) [inline deploy/olm-catalog/community-kubevirt-hyperconverged/1.13.0/manifests/kubevirt-hyperconverged-operator.v1.13.0.clusterserviceversion.yaml:12]: @nunnatsa 
is it fine we have this change ?
`make build-manifests` done that

**Comment by oshoval** [comment]
Addressed comments, lets wait for CNAO to run the tests and so on
Thanks

File: deploy/olm-catalog/community-kubevirt-hyperconverged/1.13.0/manifests/kubevirt-hyperconverged-operator.v1.13.0.clusterserviceversion.yaml
Comment (nunnatsa) [reply deploy/olm-catalog/community-kubevirt-hyperconverged/1.13.0/manifests/kubevirt-hyperconverged-operator.v1.13.0.clusterserviceversion.yaml:12]: yes. this change is expected if there additional changes in the file

File: tests/func-tests/node_placement_test.go
Comment (nunnatsa) [inline tests/func-tests/node_placement_test.go:150]: I see that it follows the pattern above, but to me it makes more sense to add the value to the map, if cnaoCR.Spec.KubevirtIpamController != nil, than this pattern.

File: deploy/olm-catalog/community-kubevirt-hyperconverged/1.13.0/manifests/kubevirt-hyperconverged-operator.v1.13.0.clusterserviceversion.yaml
Comment (oshoval) [reply deploy/olm-catalog/community-kubevirt-hyperconverged/1.13.0/manifests/kubevirt-hyperconverged-operator.v1.13.0.clusterserviceversion.yaml:12]: but there arent
unless hco00.crd.yaml in the same folder is considered as well

File: tests/func-tests/node_placement_test.go
Comment (oshoval) [reply tests/func-tests/node_placement_test.go]: updated

**Comment by oshoval** [comment]
Thanks @nunnatsa 
rebased, now it is ready please
lets see all pass

File: docs/cluster-configuration.md
Comment (nunnatsa) [inline docs/cluster-configuration.md:182]: The PR condition contains some additional important info about it. 

This file is the user guide. Please check if we need to let the user know about it here.

Comment (oshoval) [reply docs/cluster-configuration.md:182]: done

**Review by nunnatsa — APPROVED** [review]
/approve

**Comment by oshoval** [comment]
Thanks !

**Pull request merged**
